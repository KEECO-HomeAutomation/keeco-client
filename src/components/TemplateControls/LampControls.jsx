import React from 'react';
import PropTypes from 'prop-types';
import { compose, withHandlers } from 'recompose';
import { withStyles } from '@material-ui/core/styles';
import classNames from 'classnames';

import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';
import ToggleButtonGroup from '@material-ui/lab/ToggleButtonGroup';
import ToggleButton from '@material-ui/lab/ToggleButton';
import Slider from '@material-ui/lab/Slider';

const styles = theme => ({
	grid: {
		justifyContent: 'space-around'
	},
	option: {
		width: '20%'
	},
	optionText: {
		letterSpacing: '1px',
		color: theme.palette.text.secondary,
		marginTop: '12px'
	},
	action: {
		width: '80%'
	},
	slider: {
		marginTop: '24px',
		width: '95%'
	},
	thumb: {
		height: '18px',
		width: '18px',
		background: 'white',
		border: '2px solid black'
	},
	thumbR: {
		border: '2px solid red'
	},
	thumbG: {
		border: '2px solid green'
	},
	thumbB: {
		border: '2px solid blue'
	},
	track: {
		height: '6px',
		borderRadius: '1px'
	},
	trackR: {
		background: 'red'
	},
	trackG: {
		background: 'green'
	},
	trackB: {
		background: 'blue'
	},
	trackDim: {
		background: 'linear-gradient(to right, white, black)'
	}
});

export const SwitchControls = ({
	data,
	onOnChange,
	onRChange,
	onGChange,
	onBChange,
	onDimChange,
	classes
}) => {
	return (
		<Grid container className={classes.grid}>
			<Grid item className={classes.option}>
				<Typography variant="body1" className={classes.optionText}>
					On
				</Typography>
			</Grid>
			<Grid item className={classes.action}>
				<ToggleButtonGroup value={['on', 'off']} onChange={onOnChange}>
					<ToggleButton value={'on'} selected={!data.on}>Off</ToggleButton>
					<ToggleButton value={'off'} selected={data.on}>On</ToggleButton>
				</ToggleButtonGroup>
			</Grid>
			<Grid item className={classes.option}>
				<Typography variant="body1" className={classes.optionText}>
					R
				</Typography>
			</Grid>
			<Grid item className={classes.action}>
				<Slider
					min={0}
					max={255}
					onChange={onRChange}
					value={data.r || 0}
					className={classes.slider}
					classes={{
						thumb: classNames(classes.thumb, classes.thumbR),
						track: classNames(classes.track, classes.trackR)
					}}
				/>
			</Grid>
			<Grid item className={classes.option}>
				<Typography variant="body1" className={classes.optionText}>
					G
				</Typography>
			</Grid>
			<Grid item className={classes.action}>
				<Slider
					min={0}
					max={255}
					onChange={onGChange}
					value={data.g || 0}
					className={classes.slider}
					classes={{
						thumb: classNames(classes.thumb, classes.thumbG),
						track: classNames(classes.track, classes.trackG)
					}}
				/>
			</Grid>
			<Grid item className={classes.option}>
				<Typography variant="body1" className={classes.optionText}>
					B
				</Typography>
			</Grid>
			<Grid item className={classes.action}>
				<Slider
					min={0}
					max={255}
					onChange={onBChange}
					value={data.b || 0}
					className={classes.slider}
					classes={{
						thumb: classNames(classes.thumb, classes.thumbB),
						track: classNames(classes.track, classes.trackB)
					}}
				/>
			</Grid>
			<Grid item className={classes.option}>
				<Typography variant="body1" className={classes.optionText}>
					Dim
				</Typography>
			</Grid>
			<Grid item className={classes.action}>
				<Slider
					min={0}
					max={255}
					onChange={onDimChange}
					value={data.dim || 0}
					className={classes.slider}
					classes={{
						thumb: classNames(classes.thumb),
						track: classNames(classes.track, classes.trackDim)
					}}
				/>
			</Grid>
		</Grid>
	);
};

export const enhancer = compose(
	withStyles(styles),
	withHandlers({
		onOnChange: ({ onChange }) => (e, value) => onChange({ on: value[0]==='on' }),
		onRChange: ({ onChange }) => (e, value) => onChange({ r: Math.floor(value) }),
		onGChange: ({ onChange }) => (e, value) => onChange({ g: Math.floor(value) }),
		onBChange: ({ onChange }) => (e, value) => onChange({ b: Math.floor(value) }),
		onDimChange: ({ onChange }) => (e, value) => onChange({ dim: Math.floor(value) })
	})
);

export default enhancer(SwitchControls);
