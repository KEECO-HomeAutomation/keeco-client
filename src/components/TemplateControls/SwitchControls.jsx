import React from 'react';
import PropTypes from 'prop-types';
import { compose, withHandlers } from 'recompose';
import { withStyles } from '@material-ui/core/styles';

import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';
import ToggleButtonGroup from '@material-ui/lab/ToggleButtonGroup';
import ToggleButton from '@material-ui/lab/ToggleButton';

const styles = theme => ({
	grid: {
		justifyContent: 'space-around'
	},
	option: {
		width: '20%'
	},
	optionText: {
		letterSpacing: '1px',
		color: theme.palette.text.secondary,
		marginTop: '12px'
	},
	action: {
		width: '80%'
	}
});

export const SwitchControls = ({ data, onOnChange, classes }) => {
	return (
		<Grid container className={classes.grid}>
			<Grid item className={classes.option}>
				<Typography variant="body1" className={classes.optionText}>
					On
				</Typography>
			</Grid>
			<Grid item className={classes.action}>
				<ToggleButtonGroup value={['on', 'off']} onChange={onOnChange}>
					<ToggleButton value={'on'} selected={!data.on}>Off</ToggleButton>
					<ToggleButton value={'off'} selected={data.on}>On</ToggleButton>
				</ToggleButtonGroup>
			</Grid>
		</Grid>
	);
};

SwitchControls.propTypes = {
	data: PropTypes.object,
	onChange: PropTypes.func
};

export const enhancer = compose(
	withStyles(styles),
	withHandlers({
		onOnChange: ({ onChange }) => (e, value) => onChange({ on: value[0]==='on' })
	})
);

export default enhancer(SwitchControls);
