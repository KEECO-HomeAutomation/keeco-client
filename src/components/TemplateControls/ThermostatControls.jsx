import React from 'react';
import PropTypes from 'prop-types';
import { compose, withProps } from 'recompose';
import { withStyles } from '@material-ui/core/styles';

import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';
import LinearProgress from '@material-ui/core/LinearProgress';

const styles = theme => ({
	grid: {
		justifyContent: 'space-around'
	},
	option: {
		width: '20%'
	},
	optionText: {
		letterSpacing: '1px',
		color: theme.palette.text.secondary,
		marginTop: '12px'
	},
	action: {
		width: '80%'
	},
	actionText: {
		letterSpacing: '2px',
		fontSize: '1.2em'
	},
	tempBar: {
		width: '95%',
		marginLeft: '5px'
	},
	tempBarBack: {
		background:
			'linear-gradient(to right, rgba(95, 145, 226, 0.3) 0%, rgba(114, 186, 76, 0.3) 45%, rgba(114, 186, 76, 0.3) 55%, rgba(201, 145, 40, 0.3) 70%, rgba(183, 55, 44, 0.3) 100%)'
	}
});

export const SwitchControls = ({ data, processedTemp, classes }) => {
	return (
		<Grid container className={classes.grid}>
			<Grid item className={classes.option}>
				<Typography variant="body1" className={classes.optionText}>
					Temp
				</Typography>
			</Grid>
			<Grid item className={classes.action}>
				<Typography variant="body1" className={classes.actionText}>
					{data.temperature || '-'}&nbsp;&#8451;
				</Typography>
				<LinearProgress
					variant="determinate"
					value={processedTemp}
					className={classes.tempBar}
					classes={{ colorPrimary: classes.tempBarBack }}
				/>
			</Grid>
		</Grid>
	);
};

SwitchControls.propTypes = {
	data: PropTypes.object,
	onChange: PropTypes.func,
	processedTemp: PropTypes.number,
	classes: PropTypes.object
};

export const enhancer = compose(
	withStyles(styles),
	withProps(({ data }) => ({
		processedTemp: data.temperature + 30
	}))
);

export default enhancer(SwitchControls);
