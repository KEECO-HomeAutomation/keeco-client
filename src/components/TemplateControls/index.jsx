import React from 'react';
import PropTypes from 'prop-types';
import { compose, withProps } from 'recompose';
import { withStyles } from '@material-ui/core/styles';

import SwitchControls from './SwitchControls';
import LampControls from './LampControls';
import ThermostatControls from './ThermostatControls';

const styles = () => ({
	container: {
		width: '100%',
		padding: '5px'
	}
});

export const TemplateControls = ({
	Controls,
	templateData,
	onTemplateDataChange,
	classes
}) => {
	return (
		<div className={classes.container}>
			{Controls && (
				<Controls data={templateData} onChange={onTemplateDataChange} />
			)}
		</div>
	);
};

TemplateControls.propTypes = {
	template: PropTypes.string,
	templateData: PropTypes.object,
	onTemplateDataChange: PropTypes.func,
	classes: PropTypes.object,
	Controls: PropTypes.elementType
};

export const enhancer = compose(
	withStyles(styles),
	withProps(({ template }) => {
		let Controls;
		switch (template) {
			case 'switch':
				Controls = SwitchControls;
				break;
			case 'lamp':
				Controls = LampControls;
				break;
			case 'thermostat':
				Controls = ThermostatControls;
				break;
			default:
				Controls = null;
				break;
		}

		return { Controls };
	})
);

export default enhancer(TemplateControls);
