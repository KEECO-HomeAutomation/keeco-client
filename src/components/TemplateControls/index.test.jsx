/* eslint-disable react/display-name */
/* eslint-disable react/prop-types */
import React from 'react';
import { mount } from 'enzyme';

jest.mock('./SwitchControls', () => () => <p>SwitchControls</p>);
jest.mock('./LampControls', () => () => <p>LampControls</p>);
jest.mock('./ThermostatControls', () => () => <p>ThermostatControls</p>);

import TemplateControls, { TemplateControls as Base, enhancer } from './index';

import SwitchControls from './SwitchControls';
import LampControls from './LampControls';
import ThermostatControls from './ThermostatControls';

describe('<TemplateControls />', () => {
	test('Should export base and enhancer', () => {
		expect(Base).not.toBe(undefined);
		expect(enhancer).not.toBe(undefined);
	});

	test('Should render without errors', () => {
		const comp = mount(<TemplateControls />);
		expect(comp.length).toBe(1);
	});

	test('Should match snapshot', () => {
		expect(mount(<TemplateControls />)).toMatchSnapshot();
	});

	describe('Should render the needed control', () => {
		const testCases = [
			{ name: 'switch', Control: SwitchControls },
			{ name: 'lamp', Control: LampControls },
			{ name: 'thermostat', Control: ThermostatControls }
		];
		testCases.forEach(testCase => {
			test('template=' + testCase.name, () => {
				const comp = mount(<TemplateControls template={testCase.name} />);
				expect(comp).toMatchSnapshot();
				expect(comp).toContainReact(<testCase.Control />);
			});
		});
	});

	test('Should pass templateData and onTemplateDataChange to rendered controller', () => {
		const func = jest.fn();
		const comp = mount(
			<TemplateControls
				template="switch"
				templateData={{ mocked: 'data' }}
				onTemplateDataChange={func}
			/>
		);
		expect(comp.find(SwitchControls)).toHaveProp('data', { mocked: 'data' });
		expect(comp.find(SwitchControls)).toHaveProp('onChange', func);
	});
});
