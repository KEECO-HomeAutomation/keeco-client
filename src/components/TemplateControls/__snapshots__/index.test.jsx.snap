// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TemplateControls /> Should match snapshot 1`] = `
<WithStyles(withProps(TemplateControls))>
  <withProps(TemplateControls)
    classes={
      Object {
        "container": "withProps(TemplateControls)-container-1",
      }
    }
  >
    <TemplateControls
      Controls={null}
      classes={
        Object {
          "container": "withProps(TemplateControls)-container-1",
        }
      }
    >
      <div
        className="withProps(TemplateControls)-container-1"
      />
    </TemplateControls>
  </withProps(TemplateControls)>
</WithStyles(withProps(TemplateControls))>
`;

exports[`<TemplateControls /> Should render the needed control template=lamp 1`] = `
<WithStyles(withProps(TemplateControls))
  template="lamp"
>
  <withProps(TemplateControls)
    classes={
      Object {
        "container": "withProps(TemplateControls)-container-1",
      }
    }
    template="lamp"
  >
    <TemplateControls
      Controls={[Function]}
      classes={
        Object {
          "container": "withProps(TemplateControls)-container-1",
        }
      }
      template="lamp"
    >
      <div
        className="withProps(TemplateControls)-container-1"
      >
        <Component>
          <p>
            LampControls
          </p>
        </Component>
      </div>
    </TemplateControls>
  </withProps(TemplateControls)>
</WithStyles(withProps(TemplateControls))>
`;

exports[`<TemplateControls /> Should render the needed control template=switch 1`] = `
<WithStyles(withProps(TemplateControls))
  template="switch"
>
  <withProps(TemplateControls)
    classes={
      Object {
        "container": "withProps(TemplateControls)-container-1",
      }
    }
    template="switch"
  >
    <TemplateControls
      Controls={[Function]}
      classes={
        Object {
          "container": "withProps(TemplateControls)-container-1",
        }
      }
      template="switch"
    >
      <div
        className="withProps(TemplateControls)-container-1"
      >
        <Component>
          <p>
            SwitchControls
          </p>
        </Component>
      </div>
    </TemplateControls>
  </withProps(TemplateControls)>
</WithStyles(withProps(TemplateControls))>
`;

exports[`<TemplateControls /> Should render the needed control template=thermostat 1`] = `
<WithStyles(withProps(TemplateControls))
  template="thermostat"
>
  <withProps(TemplateControls)
    classes={
      Object {
        "container": "withProps(TemplateControls)-container-1",
      }
    }
    template="thermostat"
  >
    <TemplateControls
      Controls={[Function]}
      classes={
        Object {
          "container": "withProps(TemplateControls)-container-1",
        }
      }
      template="thermostat"
    >
      <div
        className="withProps(TemplateControls)-container-1"
      >
        <Component>
          <p>
            ThermostatControls
          </p>
        </Component>
      </div>
    </TemplateControls>
  </withProps(TemplateControls)>
</WithStyles(withProps(TemplateControls))>
`;
